File: js/0841-keys-and-rooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
var canVisitAllRooms = function (rooms) {
  var seen = new Array(rooms.length).fill(false);
  seen[0] = true;
  var st = [0];
  while (st.length > 0) {
    var node = st.pop();
    rooms[node].forEach((key) => {
      if (!seen[key]) {
        seen[key] = true;
        st.push(key);
      }
    });
  }

  return seen.every((val) => val == true);
};

module.exports = { canVisitAllRooms };

File: go/solutions/0841-keys-and-rooms.go
package solutions

func CanVisitAllRooms(rooms [][]int) bool {
	seen := make([]bool, len(rooms))
	seen[0] = true
	st := make([]int, 0)
	st = append(st, 0)

	for len(st) > 0 {
		node := st[len(st)-1]
		st = st[:len(st)-1]
		for _, nei := range rooms[node] {
			if !seen[nei] {
				seen[nei] = true
				st = append(st, nei)
			}
		}
	}

	for _, val := range seen {
		if !val {
			return false
		}
	}

	return true
}
